discovery_engine_core/core/src/engine.rs

```diff
@@ -72,85 +74,134 @@ where
+        Selection::new(BetaSampler, self.stacks.values())
+            .select(max_documents)
+            .await
             .map_err(|e| e.into())
     }
 
+    /// The ranker could rank the documents in a different order so we update the stacks with it.
+    async fn rank_stacks(&self) -> Result<(), Error> {
+        let errors = self
+            .stacks
+            .values()
+            .into_iter()
+            .map(|stack| async move { stack })
+            .collect::<FuturesUnordered<_>>()
+            .fold(vec![], |mut errors, stack| async move {
+                if let Err(error) = stack.write().await.rank(&self.ranker) {
+                    errors.push(Error::StackOpFailed(error));
+                }
+
+                errors
+            })
+            .await;
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::Errors(errors))
+        }
+    }
+
     /// Process the feedback about the user spending some time on a document.
-    pub fn time_logged(&mut self, time_logged: &TimeSpent) -> Result<(), Error> {
-        self.ranker.time_logged(time_logged)?;
+    pub async fn time_spent(&mut self, time_spent: &TimeSpent) -> Result<(), Error> {
+        self.ranker.time_spent(time_spent)?;
 
-        rank_stacks(self.stacks.values_mut(), &self.ranker)
+        self.rank_stacks().await
     }
 
     /// Process the feedback about the user reacting to a document.
-    pub fn user_reacted(&mut self, reacted: &UserReacted) -> Result<(), Error> {
-        let stack = self
-            .stacks
-            .get_mut(&reacted.stack_id)
-            .ok_or(Error::InvalidStackId(reacted.stack_id))?;
-
-        stack.update_relevance(reacted.reaction);
+    pub async fn user_reacted(&mut self, reacted: &UserReacted) -> Result<(), Error> {
+        self.stacks
+            .get(&reacted.stack_id)
+            .ok_or(Error::InvalidStackId(reacted.stack_id))?
+            .write()
+            .await
+            .update_relevance(reacted.reaction);
 
         self.ranker.user_reacted(reacted)?;
 
-        rank_stacks(self.stacks.values_mut(), &self.ranker)
+        self.rank_stacks().await
     }
-}
-
-/// The ranker could rank the documents in a different order so we update the stacks with it.
-fn rank_stacks<'a, R: Ranker>(
-    stacks: impl Iterator<Item = &'a mut Stack>,
-    ranker: &R,
-) -> Result<(), Error> {
-    let errors = stacks.fold(vec![], |mut errors, stack| {
-        if let Err(e) = stack.rank(ranker).map_err(Error::StackOpFailed) {
-            errors.push(e);
-        }
-
-        errors
-    });
 
-    if errors.is_empty() {
-        Ok(())
-    } else {
-        Err(Error::Errors(errors))
+    /// Updates the stacks with data related to the top key phrases of the current data.
+    #[allow(dead_code)]
+    async fn update_stacks(&mut self, top: usize) -> Result<(), Error> {
+        let key_phrases = &self.ranker.select_top_key_phrases(top);
+        let ranker = &self.ranker;
```

Comment:
since we now require `&mut` we could use `while let`

```rust
        let key_phrases = &self.ranker.select_top_key_phrases(top);
        let ranker = &mut self.ranker;

        while let Some(stack) = self.stacks.values().into_iter().next() {
            let (stack, documents) = stack
                .read()
                .await
                .ops
                .new_items(key_phrases, ranker)
                .await
                .map(|documents| (stack, documents))?;

            stack
                .write()
                .await
                .update(&documents, ranker)
                .map_err(GenericError::from)?
        }

        Ok(())
```

Reply:
this would make the loop synchronous again. currently we collect the "new items future" into the `FuturesUnordered` and run the async try_collect loop on those.

Reply:
my bad, now i get what you mean, we can't pass in `ranker` anyways anymore due to the owned reference. i'll check if we can work around this asynchronously in some way, otherwise i'll follow your suggestion.

Reply:
I'm not so sure if we're going to benefit that much from the async stuff either. Since almost every engine method requires `&mut`, we will only be able to process the engine calls sequentially and not concurrently. expect we are wrapping the engine in a mutex. But this is something that we can tackle later.